libraries/lib-stretching-sequence/StretchingSequence.h

```diff
@@ -0,0 +1,83 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  StretchingSequence.h
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#pragma once
+
+#include "AudioIOSequences.h"
+#include "PlaybackDirection.h"
+
+#include <memory>
+#include <optional>
+
+class AudioSegment;
+class AudioSegmentFactoryInterface;
+class ClipInterface;
+using ClipHolders = std::vector<std::shared_ptr<ClipInterface>>;
+using ClipConstHolders = std::vector<std::shared_ptr<const ClipInterface>>;
+
+// For now this class assumes forward reading, which will be sufficient for the
+// first goal of allowing export and rendering.
+class STRETCHING_SEQUENCE_API StretchingSequence final : public PlayableSequence
+{
+public:
+   static std::shared_ptr<StretchingSequence>
+   Create(const PlayableSequence&, const ClipConstHolders& clips);
+   static std::shared_ptr<StretchingSequence>
+   Create(const PlayableSequence&, const ClipHolders& clips);
+
+   StretchingSequence(
+      const PlayableSequence&, int sampleRate, size_t numChannels,
+      std::unique_ptr<AudioSegmentFactoryInterface>);
```

Comment:
If the constructor is not meant for public use, but instead only the `Create` factories -- maybe use the `CreateToken` trick as in `EffectStage`.  The constructor has to be public, but `Create` needs make-shared so `std` functions have to call the constructor.  But the trick is a way to restrict the usage without the `std` functions needing to be `friend`s.

Reply:
It's ok, it can be used publicly, it's just that clients usually don't want to bother about injecting a factory and are happy with the default one. One test case, though, needs it.
But that's a nice trick, never thought about it.